dv/verilator/memutil/cpp/verilator_memutil.h

```diff
@@ -0,0 +1,103 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef VERILATOR_MEMUTIL_H_
+#define VERILATOR_MEMUTIL_H_
+
+#include <vltstd/svdpi.h>
+
+#include <map>
+#include <string>
+
+enum MemImageType {
+  kMemImageUnknown = 0,
+  kMemImageElf,
+  kMemImageVmem,
+};
+
+struct MemArea {
+  std::string name;      // Unique identifier
+  std::string location;  // Design scope location
+};
+
+/**
+ * Provide various memory loading utilities for Verilator simulations
+ *
+ * These utilities require the corresponding DPI functions:
+ * simutil_verilator_memload()
+ * simutil_verilator_set_mem()
+ * to be defined somewhere as SystemVerilog functions.
+ */
+class VerilatorMemUtil {
+ public:
+  /**
+   * Register a memory as instantiated by generic ram
+   *
+   * The |name| must be a unique identifier. The function will return false
+   * if |name| is already used. |location| is the path to the scope of the
+   * instantiated memory, which needs to support the DPI-C interfaces
+   * 'simutil_verilator_memload' and 'simutil_verilator_set_mem' used for
+   * 'vmem' and 'elf' files, respectively.
+   *
+   * Memories must be registered before command arguments are parsed by
+   * ParseCommandArgs() in order for them to be known.
+   */
+  bool RegisterMemoryArea(const std::string name, const std::string location);
+
+  /**
+   * Parse command line arguments
+   *
+   * This removes all recognized command-line arguments from argc/argv.
+   *
+   * The return value of this method indicates if the program should exit with
+   * retcode: if this method returns true, do *not* exit; if it returns *false*,
+   * do exit.
+   */
+  bool ParseCommandArgs(int argc, char **argv, bool &exit_app);
```

Comment:
exit_app isn't documented.

Looking at the implementation seems it used to indicate you should exit because it's been asked to just list memory regions (i.e. no error has occurred) but the two bools (one return value, one exit now but no error) feels a bit awkward.

Maybe return bool for exit/no exit and pass by reference integer retcode to return? Or have both bools pass by reference so it's more obvious what's going on (one called error, the other exit_app).

Reply:
We went around in circles on this one in the previous PRs, and that's the best solution we could come up with. bool return code is the standard convention for all functions in this module. It indicates if the function was successful or not. But even a successful call can mean you want to end the program. That's what exit_app is for.

Reply:
I'll fix the comment here and in simctrl